From 14381ec246f67c2e3b8ae0c569fc41d214bfe648 Mon Sep 17 00:00:00 2001
From: linmin <linmin@eswincomputing.com>
Date: Thu, 30 May 2024 09:57:16 +0800
Subject: [PATCH 064/109] WIN2030-15279:fix(hf_power_process.c):Fix failure of
 getting board power occasionally

Changelogs:
1.After running for along time, it might fail to get board power through I2C.
  The log shows "I2Cx_read_Error(88) reg 3; status 1"
  Added taskENTER_CRITICAL()/taskEXIT_CRITICAL() during I2C operation to fix this issue.

Change-Id: I5ccbf92ef801088c2ca05f1849b5b29c9699b3ab
---
 Core/Src/hf_power_process.c | 46 ++++++++++++++++++++++++++-----------
 1 file changed, 32 insertions(+), 14 deletions(-)

diff --git a/Core/Src/hf_power_process.c b/Core/Src/hf_power_process.c
index 1705953..b0e21a4 100644
--- a/Core/Src/hf_power_process.c
+++ b/Core/Src/hf_power_process.c
@@ -426,34 +426,42 @@ int get_board_power(uint32_t *volt, uint32_t *curr, uint32_t *power)
 	uint32_t reg_curr = 0x0;
 	int ret = 0;
 	static int is_first = 1;
+	struct rtc_time_t time = {0};
 
+	es_get_rtc_time(&time);
+
+	taskENTER_CRITICAL();
 	if (1 == is_first)
 	{
 		/*Write failure, not find wrong reason*/
 		ret = ina226_init();
 		if (ret)
 		{
-			return -1;
+			ret = -1;
+			goto out;
 		}
 		is_first = 0;
 	}
 	ret = hf_i2c_reg_read_block(&hi2c3, INA226_12V_ADDR, INA2XX_BUS_VOLTAGE, (uint8_t *)&reg_bus, 2);
 	if (ret)
 	{
-		printf("get board volt error\n");
-		return -1;
+		printf("get board volt error(%02u:%02u:%02u CST)\n", time.Hours, time.Minutes, time.Seconds);
+		ret = -1;
+		goto out;
 	}
 	ret = hf_i2c_reg_read_block(&hi2c3, INA226_12V_ADDR, INA2XX_POWER, (uint8_t *)&reg_power, 2);
 	if (ret)
 	{
-		printf("get board power error\n");
-		return -1;
+		printf("get board power error(%02u:%02u:%02u CST)\n", time.Hours, time.Minutes, time.Seconds);
+		ret = -1;
+		goto out;
 	}
 	ret = hf_i2c_reg_read_block(&hi2c3, INA226_12V_ADDR, INA2XX_CURRENT, (uint8_t *)&reg_curr, 2);
 	if (ret)
 	{
-		printf("get board current error\n");
-		return -1;
+		printf("get board current error(%02u:%02u:%02u CST)\n", time.Hours, time.Minutes, time.Seconds);
+		ret = -1;
+		goto out;
 	}
 	reg_bus = SWAP16(reg_bus);
 	reg_curr = SWAP16(reg_curr);
@@ -466,7 +474,9 @@ int get_board_power(uint32_t *volt, uint32_t *curr, uint32_t *power)
 	*curr = reg_curr * INA226_CURRENT_LSB;
 	*curr = DIV_ROUND_CLOSEST(*curr, 1000);
 
-	return 0;
+out:
+	taskEXIT_CRITICAL();
+	return ret;
 }
 
 
@@ -481,6 +491,7 @@ int get_som_power(uint32_t *volt, uint32_t *curr, uint32_t *power)
 	uint8_t is_neg = 0;
 	uint8_t ctrl_value = 0x8;
 
+	taskENTER_CRITICAL();
 	if (1 == is_first)
 	{
 		/*Write failure, not find wrong reason*/
@@ -488,7 +499,8 @@ int get_som_power(uint32_t *volt, uint32_t *curr, uint32_t *power)
 
 		if (ret)
 		{
-			return -1;
+			ret = -1;
+			goto out;
 		}
 		is_first = 0;
 	}
@@ -499,13 +511,15 @@ int get_som_power(uint32_t *volt, uint32_t *curr, uint32_t *power)
 	if (ret)
 	{
 		printf("get PAC1934 act_val error\n");
-		return -1;
+		ret = -1;
+		goto out;
 	}
 	ret = hf_i2c_reg_read_block(&hi2c3, PAC1934_ADDR, PAC193X_CMD_VBUS1, (uint8_t *)&reg_bus, 2);
 	if (ret)
 	{
 		printf("get PAC1934 voltage error\n");
-		return -1;
+		ret = -1;
+		goto out;
 	}
 	reg_bus = reg_bus & 0XFFFF;
 	reg_bus = SWAP16(reg_bus);
@@ -523,7 +537,8 @@ int get_som_power(uint32_t *volt, uint32_t *curr, uint32_t *power)
 	if (ret)
 	{
 		printf("get PAC1934 current error\n");
-		return -1;
+		ret = -1;
+		goto out;
 	}
 	reg_curr = reg_curr & 0XFFFF;
 	if (0x1 == ((act_val >> 7) & 0x1))
@@ -540,7 +555,8 @@ int get_som_power(uint32_t *volt, uint32_t *curr, uint32_t *power)
 	if (ret)
 	{
 		printf("get PAC1934 power error\n");
-		return -1;
+		ret = -1;
+		goto out;
 	}
 	reg_power = reg_power >> 4;
 	if (1 == is_neg)
@@ -552,7 +568,9 @@ int get_som_power(uint32_t *volt, uint32_t *curr, uint32_t *power)
 		*power = reg_power * PAC193X_COSTANT_PWR_M / (PAC193X_SHUNT_RESISTOR_M * 268435456ULL);
 	}
 
-	return 0;
+out:
+	taskEXIT_CRITICAL();
+	return ret;
 }
 
 power_switch_t get_som_power_state(void)
-- 
2.25.1

